enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+  - abhinavdahiya
+  - joelspeed
+  - smarterclayton
+  - derekwaynecarr
+approvers:
+  - TBD
+creation-date: 2020-04-02
+last-updated: yyyy-mm-dd
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - "https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md"
+  - "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/cluster-etcd-operator.md"
+  - "https://github.com/openshift/enhancements/blob/master/enhancements/etcd/disaster-recovery-with-ceo.md"
+  - "https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md"
+  - "https://github.com/openshift/cluster-kube-scheduler-operator"
+  - "https://github.com/openshift/cluster-kube-controller-manager-operator"
+  - "https://github.com/openshift/cluster-openshift-controller-manager-operator"
+  - "https://github.com/openshift/cluster-kube-apiserver-operator"
+  - "https://github.com/openshift/cluster-openshift-apiserver-operator"
+---
+
+# Single entity Control Plane
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For clarity in this doc we set the definition for "Control Plane" as "The collection of stateless and stateful processes which enable a Kubernetes cluster to meet minimum operational requirements". This includes: kube-apiserver, kube-controller-manager, kube-scheduler, kubelet and Etcd.
+
+This proposal outlines a solution for declaratively managing as a single entity the compute resources that host the OCP Control Plane components. It introduces scaling and self-healing capabilities for this compute resources while honouring inviolable Etcd expectations and with out disrupting the lifecycle of Control plane components.
+
+## Motivation
+
+The Control Plane is the most critical and sensitive entity of a running cluster. Today OCP Control Plane instances are "pets" and therefore fragile. There are multiple scenarios where adjusting the compute capacity which is backing the Control Plane components might be desirable either for resizing or repairing. 
+Currently there is nothing that automates or eases this task. The steps for the Control Plane to be resized in any manner or to recover from a tolerable failure are completely manual. Different teams are following different SOP documents scattered around with manual steps resulting in lost of information, confusion and extra efforts for engineers and users.
+
+### Goals
+
+- To have a declarative mechanism to manage the Control Plane as a single entity.
+- To support declarative safe horizontal scaling towards an odd number replicas for the control plane compute resources.
```

Comment:
Having an even number of replicas is not inherently bad.
You don't achieve any additional failure tolerance by adding the last node, but maybe there is a performance benefit.
So perhaps it is not beneficial to enforce in code.

Reply:
Is it fair to say [etcd docs](https://github.com/etcd-io/etcd/blob/master/Documentation/faq.md) argue strongly that an even number of replicas is inherently bad?

> Although adding a node to an odd-sized cluster appears better since there are more machines, the fault tolerance is worse since exactly the same number of nodes may fail without losing quorum but there are more nodes that can fail.

> during a network partition, an odd number of members guarantees that there will always be a majority partition that can continue to operate and be the source of truth when the partition ends

Reply:
I'm sure they do, if one knows nothing about a given setup, then "odd" is the most reliable advice to give.

However it is possible to construct scenarios where 2*N nodes is less vulnerable to outages than 2*N + 1 by varying how the machines are organised/connected (ie. the number of racks).

So advise an odd number of nodes: absolutely
Enforce an odd number of nodes: my vote is no